schema_registry/serializers/message_serializer.py

```diff
@@ -62,51 +63,35 @@ def _serializer_schema_type(self) -> str:
         ...
 
     @abstractmethod
-    def _get_encoder_func(self, schema: BaseSchema) -> typing.Callable:
+    def _get_serializer_func(self, schema: BaseSchema) -> typing.Callable:
         ...
 
     @abstractmethod
-    def _get_decoder_func(self, payload: ContextStringIO, writer_schema: BaseSchema) -> typing.Callable:
+    def _get_deserializer_func(self, payload: ContextStringIO, writer_schema: BaseSchema) -> typing.Callable:
         ...
 
-    def encode_record_with_schema(self, subject: str, schema: typing.Union[BaseSchema], record: dict) -> bytes:
-        """
-        Given a parsed avro schema, encode a record for the given subject.
-        The record is expected to be a dictionary.
-        The schema is registered with the subject of 'topic-value'
-        Args:
-            subject (str): Subject name
-            schema (avro.schema.RecordSchema): Avro Schema
-            record (dict): An object to serialize
-        Returns:
-            bytes: Encoded record with schema ID as bytes
-        """
-        # Try to register the schema
-        schema_id = self.schemaregistry_client.register(subject, schema, schema_type=self._serializer_schema_type)
-
-        # cache writer
-        if not self.id_to_writers.get(schema_id):
-            self.id_to_writers[schema_id] = self._get_encoder_func(schema)
+    @functools.singledispatchmethod
+    def serialize(self, schema_or_id: typing.Any, record: dict, *_: typing.Any) -> bytes:
```

Comment:
the way `singledispatchmethod` works is that the type of the first argument (after `self`) is inspected and then it finds the registered method with the matching type.
I rearranged the arguments to _serialize_with_schema_and_subject so that the most import argument was the determining factor.
If you pass in an integer, it goes to the _serialize_by_id, if you pass in an instance of BaseSchema, it goes to _serialize_by_schema_and_subject.